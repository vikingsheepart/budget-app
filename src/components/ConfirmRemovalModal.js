import React from 'react';
import PropTypes from 'prop-types';
import Modal from 'react-modal';

Modal.defaultStyles.overlay.backgroundColor = 'translucent blue';

const ConfirmRemovalModal = props => (
  <Modal
    className="modal"
    contentLabel="Remove Expense"
    isOpen={props.isModalOpen}
    onRequestClose={props.onToggleModal} // esc or click outside to close modal
    closeTimeoutMS={200}
  >
    <h2 className="modal__title">Removing: {props.expense.description}</h2>
    <p className="modal__body">
      All the information will be deleted from the database.<br />
      This action cannot be undone.
    </p>
    <div className="modal__buttons">
      <button className="button" onClick={props.onToggleModal}>Keep Expense</button>
      <button className="button button--danger" onClick={props.onRemove}>Delete Expense</button>
    </div>
  </Modal>
);

ConfirmRemovalModal.propTypes = {
  expense: PropTypes.shape().isRequired,
  isModalOpen: PropTypes.bool.isRequired,
  onToggleModal: PropTypes.func.isRequired,
  onRemove: PropTypes.func.isRequired
};

export default ConfirmRemovalModal;
