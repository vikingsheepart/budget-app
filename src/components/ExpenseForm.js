import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    // props.expense is passed from EditExpensePage
    this.state = {
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      note: props.expense ? props.expense.note : '',
      calendarFocused: false,
      errorDescription: '',
      errorAmount: ''
    };
  }
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description, errorDescription: '' }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    // allow users to delete the whole value/leave input empty
    // allow input to begin with numbers with at most 2 decimal places
    // no letters allowed
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount, errorAmount: '' }));
    }
  };
  onDateChange = (createdAt) => {
    // prevent users from deleting value
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  }
  onSubmit = (e) => {
    e.preventDefault();
    // input validation
    if (!this.state.description || !this.state.amount) {
      if (!this.state.description) {
        this.setState(() => ({ errorDescription: 'Please provide description' }));

      }
      if (!this.state.amount) {
        this.setState(() => ({ errorAmount: 'Please provide amount' }));
      } 
    } else {
      // Clear the error
      this.setState(() => ({
        errorDescription: '',
        errorAmount: ''
      }));
      // Pass newly created expense to AddExpensePage and EditExpensePage
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100, // parse string into number (unit: penny)
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      })
    }
  }
  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.errorDescription && <p className="form__error">{this.state.errorDescription}</p>}
        <input
          className={this.state.errorDescription ? "text-input text-input__error" : "text-input"}
          type="text"
          placeholder="Description"
          value={this.state.description}
          autoFocus
          onChange={this.onDescriptionChange}
        />
        {this.state.errorAmount && <p className="form__error">{this.state.errorAmount}</p>}
        <input
          className={this.state.errorAmount ? "text-input text-input__error" : "text-input"}
          type="text"
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.onAmountChange}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false} // allow selecting past days
          showDefaultInputIcon
          hideKeyboardShortcutsPanel
          block
        />
        <textarea
          className="textarea"
          placeholder="Add a note for your expense (optional)"
          value={this.state.note}
          onChange={this.onNoteChange}
        >
        </textarea>
        <div>
          <button className="button">Save Expense</button>
        </div>
      </form>
    );
  }
}
