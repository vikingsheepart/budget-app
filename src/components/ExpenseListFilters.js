import React from 'react';
import PropTypes from 'prop-types';
import momentPropTypes from 'react-moment-proptypes';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  };
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };
  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              className="text-input"
              type="text"
              value={this.props.filters.text}
              placeholder="Search Expenses"
              onChange={this.onTextChange}
            />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              startDateId="startDate" // for prop validation
              endDateId="endDate" // for prop validation
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              showClearDates={true}
              numberOfMonths={1}
              isOutsideRange={() => false} // allow selecting past days
              showDefaultInputIcon
              hideKeyboardShortcutsPanel
            />
          </div>
        </div>
      </div>
    );
  }
};

const mapStateToProps = state => ({
  filters: state.filters
});

const mapDispatchToProps = dispatch => ({
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate)),
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount())
});

ExpenseListFilters.propTypes = {
  filters: PropTypes.shape({
    text: PropTypes.string.isRequired,
    sortBy: PropTypes.string.isRequired,
    startDate: momentPropTypes.momentObj,
    endDate: momentPropTypes.momentObj
  }).isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
