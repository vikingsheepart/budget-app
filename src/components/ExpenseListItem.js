import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({ id, description, amount, createdAt, note }) => (
  <Link className="list-item" to={`/edit/${id}`}>
    <div>
      <h3 className="list-item__title">{description}</h3>
      <span className="list-item__sub-title">{moment(createdAt).format('MMMM Do, YYYY')}</span>
    </div>
    <span className="list-item__note">{note}</span>
    <h3 className="list-item__amount">{numeral(amount / 100).format('$0,00.00')}</h3>
  </Link>
);

ExpenseListItem.propTypes = {
  id: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  amount: PropTypes.string.isRequired,
  createdAt: PropTypes.number.isRequired,
  note: PropTypes.string.isRequired
};

export default ExpenseListItem;
