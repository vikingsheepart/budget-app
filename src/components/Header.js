import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';

export const Header = ({ onLogout }) => (
  <header className="header">
    <div className="content-container">
      <div className="header__content">
        <Link className="header__title" to="/dashboard">
          <h1>Budget App</h1>
        </Link>
        <button className="button button--no-bg-sd" onClick={onLogout}>Logout</button>
      </div>
    </div>
  </header>
);

const mapDispatchToProps = dispatch => ({
  onLogout: () => dispatch(startLogout())
});

Header.propTypes = {
  onLogout: PropTypes.func.isRequired
};

export default connect(undefined, mapDispatchToProps)(Header);
