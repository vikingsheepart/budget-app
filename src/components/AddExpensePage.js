import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

// export class for testing
export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    // this.props.dispatch(addExpense(expense));
    this.props.startAddExpense(expense);
    // automate redirect after submitting the form
    this.props.history.push('/');
  }
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm onSubmit={this.onSubmit} />
        </div>
      </div>
    )
  }
}

// return an object as a Redux action creator
const mapDispatchToProps = dispatch => ({
  startAddExpense: expense => dispatch(startAddExpense(expense))
});

AddExpensePage.propTypes = {
  startAddExpense: PropTypes.func.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired
};

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
