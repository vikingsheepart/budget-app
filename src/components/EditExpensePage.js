import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import ConfirmRemovalModal from './ConfirmRemovalModal';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

// Refactor EditExpensePage to be a class based component for better performance
// check https://goo.gl/EYjYLo
// export class for testing
export class EditExpensePage extends React.Component {
  state = {
    isModalOpen: false
  };
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };
  onToggleModal = () => {
    this.setState(() => ({ isModalOpen: !this.state.isModalOpen}));
  };
  onRemove = () => {
    this.setState(() => ({ isModalOpen: false }));
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            // this.props.expense is the state.expense passed from mapStateToProps
            expense={this.props.expense}
            onSubmit={this.onSubmit}
          />
          <button className="button button--secondary" onClick={this.onToggleModal}>
            Remove Expense
          </button>
        </div>
        <ConfirmRemovalModal
          expense={this.props.expense}
          isModalOpen={this.state.isModalOpen}
          onToggleModal={this.onToggleModal}
          onRemove={this.onRemove}
        />
      </div>
    )
  }
}

const mapStateToProps = (state, props) => ({
  // props.match.params.id is expense.id passed from ExpenseListItem <- ExpenseList
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = dispatch => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: ({ id }) => dispatch(startRemoveExpense({ id }))
});

EditExpensePage.propTypes = {
  expense: PropTypes.shape().isRequired,
  startEditExpense: PropTypes.func.isRequired,
  startRemoveExpense: PropTypes.func.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);

// ======================= Functional Component ============================
// const EditExpensePage = props => (
//   <div>
//     <ExpenseForm
//       // props.expense is the state.expense passed from mapStateToProps
//       expense={props.expense}
//       onSubmit={(expense) => {
//         // Dispatch the action to edit the expense
//         props.dispatch(editExpense(props.expense.id, expense));
//         // redirect to the dashboard
//         props.history.push('/');
//       }}
//     />
//     <button
//       onClick={() => {
//         props.dispatch(removeExpense({ id: props.expense.id }));
//         props.history.push('/');
//       }}
//     >
//       Remove
//     </button>
//   </div>
// );
// ======================= End Functional Component ============================
