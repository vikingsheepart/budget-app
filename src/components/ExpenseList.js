import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

// export ExpenseList for testing
export const ExpenseList = props => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>
    <div className="list-body">
      {
        // props.expenses is the state passed from mapStateToProps
        props.expenses.length === 0 ? (
          <div className="list-item list-item--message">
            <span>No expenses</span>
          </div>
        ) : (
          props.expenses.map(expense => (
            <ExpenseListItem key={expense.id} {...expense} />
          ))
        )
      }
    </div>
  </div>
);

const mapStateToProps = state => ({
  // state = store that is being passed from app.js
  expenses: selectExpenses(state.expenses, state.filters)
});

ExpenseList.propTypes = {
  expenses: PropTypes.arrayOf(PropTypes.object).isRequired
};

export default connect(mapStateToProps)(ExpenseList);
