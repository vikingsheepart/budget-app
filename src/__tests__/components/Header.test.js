import React from 'react';
import { shallow } from 'enzyme';
import { Header } from '../../components/Header';

test('should render Header correctly', () => {
  const wrapper = shallow(<Header onLogout={() => { }} />);
  expect(wrapper).toMatchSnapshot();
});

test('should call onLogout on button click', () => {
  const onLogoutSpy = jest.fn();
  const wrapper = shallow(<Header onLogout={onLogoutSpy} />);
  wrapper.find('button').simulate('click');
  expect(onLogoutSpy).toHaveBeenCalled();
});
