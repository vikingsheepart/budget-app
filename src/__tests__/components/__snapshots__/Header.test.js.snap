// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly 1`] = `
<header
  className="header"
>
  <div
    className="content-container"
  >
    <div
      className="header__content"
    >
      <Link
        className="header__title"
        replace={false}
        to="/dashboard"
      >
        <h1>
          Budget App
        </h1>
      </Link>
      <button
        className="button button--no-bg-sd"
        onClick={[Function]}
      >
        Logout
      </button>
    </div>
  </div>
</header>
`;
